zerver/actions/message_send.py

```diff
@@ -1268,6 +1270,42 @@ class UserData(TypedDict):
     return sent_message_results
 
 
+def update_direct_message_group_stats(send_message_requests: Sequence[SendMessageRequest]) -> None:
+    """Update first_message, last_message, and total_messages for DirectMessageGroups."""
+    # Sort send_requests by message ID to ensure correct ordering when
+    # multiple messages are sent to the same DM group
+    dm_group_send_requests = [
+        send_request
+        for send_request in send_message_requests
+        if send_request.message.recipient.type == Recipient.DIRECT_MESSAGE_GROUP
+    ]
+    dm_group_send_requests.sort(key=lambda req: req.message.id)
+
+    if not dm_group_send_requests:
+        return
+
+    direct_message_group_ids = [
+        send_request.message.recipient.type_id for send_request in dm_group_send_requests
+    ]
+
+    # Fetch DirectMessageGroups with explicit ordering to prevent deadlocks
+    direct_message_groups = DirectMessageGroup.objects.filter(
+        id__in=direct_message_group_ids
+    ).order_by("id")
+    direct_message_group_map = {dmg.id: dmg for dmg in direct_message_groups}
+
+    for send_request in dm_group_send_requests:
+        direct_message_group = direct_message_group_map[send_request.message.recipient.type_id]
+        if direct_message_group.first_message_id is None:
+            direct_message_group.first_message_id = send_request.message.id
+        direct_message_group.last_message_id = send_request.message.id
+        direct_message_group.total_messages += 1
+
+    DirectMessageGroup.objects.bulk_update(
+        direct_message_groups, ["first_message_id", "last_message_id", "total_messages"]
+    )
```

Comment:
This adds two database queries to each DM message send, which will add some latency. I imagine it could be fairly easily reduced to just one by doing this logic in SQL, rather than pulling the `DirectMessageGroup` down in Python here. Though the better approach might be for the `DirectMessageGroup` to have been already fetched in the `check_message` code path, and thus be available here -- we'll be wanting to fetch it as part of constructing a `MessageRequest` to check any properties that the group has.

I believe `recipient_for_user_profiles` is already fetching it, so adjusting that function to also return the `DirectMessageGroup`, if any, would work (and `None` for `direct_message_group` if it's and old-style DM)